### ISSUE TITLE
[Feature] Enrichment and Prune Rosenbluth Method (PERM) -Monte Carlo-

### ISSUE THREAD

**Issue by phcerdan** [glotzerlab/hoomd-blue#445]
## Description
Add Enrichment and Prune Rosenbluth Method (PERM) to Monte Carlo module.

Successful method to simulate polymers (especially single chains) with MC. See next section for references.

The sampling changes, from Gibbs sampling (Metropolis), where all configurations have the same weight for the partition function. Metropolis achieves perfect importance sampling, at the cost of highly correlated trials. PERM tries a different strategy, a compromise between importance sampling and independence. PERM uses prune and enrichment changing the weights of these configurations.

## Additional context
[Original Paper](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.56.3682) - Pruned-enriched Rosenbluth method: Simulations of theta polymers of chain length up to 1 000 000..., P. Grassberger
[Recent Review](http://arxiv.org/abs/1107.1214) - A review of Monte Carlo simulations of polymers with PERM, H.P. Hsu and P. Grassberger

[basnijholt/Markov-chain-Monte-Carlo-polymer-growth](https://github.com/basnijholt/Markov-chain-Monte-Carlo-polymer-growth/blob/master/PERM%20-%20Rosenbluth.ipynb) - Prototype implementation in python.

I haven't found any open implementation, the algorithm seems related to genetic algorithms, with the enrich/prune step.

## Developer

I develop every day in C++, but not familiar with hoomd codebase. But with some guidance I might be able to help.

The change would involve to tweak the `update()` in the integrator. For example, here: https://github.com/glotzerlab/hoomd-blue/blob/master/hoomd/hpmc/IntegratorHPMCMono.h#L812

Edit: I found there was a PR in Hoomd involving Rosenbluth steps here: [95/slip-springs-functionality-to-mimic](https://bitbucket.org/glotzer/hoomd-blue/pull-requests/95/slip-springs-functionality-to-mimic). It might be helpful for this feature. Is there any list of plugins of hoomd, to see if that PR was converted into one?

**Comment by joaander** [comment]
Thanks for the suggestion. What makes HOOMD a good fit for PERM?

**Comment by phcerdan** [comment]
Thanks for the reply @joaander, I think PERM is a good way to generate polymers, and it involves Montecarlo methods, which HOOMD knows about.
I guess the description of the project in github was the reason I opened this issue:

>  Molecular dynamics and Monte Carlo soft matter simulation on GPUs

Meanwhile, I have created my own PERM implementation:
https://github.com/phcerdan/perm-montecarlo/

that is good for my purposes for now, but it would be great to have a GPU implementation as well, which again is another area of expertise of HOOMD!

Closing for now, if somebody with some time and experience with HOOMD codebase wants to re-open, I would be glad to help.
